core/ast_node.js

```diff
@@ -0,0 +1,657 @@
+  } else if (connection.type === Blockly.PREVIOUS_STATEMENT) {
+    astNode = new Blockly.ASTNode(Blockly.ASTNode.types.PREVIOUS, connection);
+  }
+  return astNode;
+};
+
+/**
+ * Creates an ast node from an input. Stores the connection of the input instead
+ * of the actual input.
+ * @param {?Blockly.Input} input The input used to create an ast node.
+ * @return {?Blockly.ASTNode} The ast node for the input.
+ */
+Blockly.ASTNode.createInputNode = function(input) {
+  if (!input){return;}
+  return new Blockly.ASTNode(Blockly.ASTNode.types.INPUT, input);
+};
+
+/**
+ * Creates an ast node for a block.
+ * @param {?Blockly.Block} block The block used to create an ast node.
+ * @return {?Blockly.ASTNode} The ast node for the block.
+ */
+Blockly.ASTNode.createBlockNode = function(block) {
+  if (!block){return;}
+  return new Blockly.ASTNode(Blockly.ASTNode.types.BLOCK, block);
+};
+
+/**
+ * Create an ast node of type stack.
+ * @param {?Blockly.Block} topBlock The top block of a stack.
+ * @return {?Blockly.ASTNode} [description]
+ */
+Blockly.ASTNode.createStackNode = function(topBlock) {
+  if (!topBlock){return;}
+  return new Blockly.ASTNode(Blockly.ASTNode.types.STACK, topBlock);
+};
+
+/**
+ * Creates an ast node for a workpsace.
+ * @param {?Blockly.Workspace} workspace The workspace that we are on.
+ * @param {?goog.math.Coordinate} wsCoordinate The position on the workspace for
+ *     this node.
+ * @return {?Blockly.ASTNode} The node for a workspace.
+ */
+Blockly.ASTNode.createWorkspaceNode = function(workspace, wsCoordinate) {
+  if (!workspace || !wsCoordinate){return;}
+
+  var params = {
+    "wsCoordinate": wsCoordinate
+  };
+  return new Blockly.ASTNode(
+      Blockly.ASTNode.types.WORKSPACE, workspace, params);
+};
+
+/**
+ * Object holding different types for an ASTNode.
+ */
+Blockly.ASTNode.types = {
+  FIELD: 'field',
+  BLOCK: 'block',
+  INPUT: 'input',
+  OUTPUT: 'output',
+  NEXT: 'next',
+  PREVIOUS: 'previous',
+  STACK: 'stack',
+  WORKSPACE: 'workspace'
+};
+
+/**
+ * The amount to move the workspace coordinate.
+ * @type {number}
+ * @private
+ */
+Blockly.ASTNode.wsMove_ = 10;
+
+/**
+ * The default position on the worskpace when starting from a block.
+ * @type {number}
+ * @private
+ */
+Blockly.ASTNode.defaultPosition_ = new goog.math.Coordinate(100,100);
+
+/**
+ * Parse the optional parameters.
+ * @param {?Object} params The user specified parameters.
+ * @private
+ */
+Blockly.ASTNode.prototype.processParams_ = function(params){
+  if (!params) {return;}
+
+  if (params['wsCoordinate']) {
+    this.wsCoordinate_ = params['wsCoordinate'];
+  }
+};
+
+/**
+ * Gets the current location of the cursor.
+ * @return {!Blockly.Field|Blockly.Connection|Blockly.Block|Blockly.Workspace}
+ * The current field, connection, workspace, or block the cursor is on.
+ */
+Blockly.ASTNode.prototype.getLocation = function() {
+  return this.location_;
+};
+
+/**
+ * The type of the current location.
```

Comment:
add "One of Blockly.ASTNode.types"

Reply:
Done. Changes can be found in cursor_ast_update branch.